editor/src/core/es-modules/package-manager/module-resolution.ts

```diff
@@ -1,180 +1,261 @@
+  if (isResolveSuccess(jsonFileResult) && isEsCodeFile(jsonFileResult.success.file)) {
+    // Parse X/package.json, and look for "browser" or "main" field
+    const mainEntryPath = processPackageJson(jsonFileResult.success.file.fileContents)
 
-function fallbackLookup(
-  lookupFn: FileLookupFn,
-  path: Array<string>,
-  direct: boolean,
-): FileLookupResult {
-  const asIsPath = normalizePath(path)
-  const pathToFile = dropLast(asIsPath)
-  const lastPart = last(asIsPath)
-  if (lastPart == null) {
-    return resolveNotPresent
-  } else {
-    const asIsResult = lookupFn(asIsPath, direct)
-    if (isResolveSuccess(asIsResult)) {
-      return asIsResult
-    }
+    if (mainEntryPath != null) {
+      // let M = X + (json main field)
+      const mainEntryPathParts = [...normalisedPathParts, ...getPartsFromPath(mainEntryPath)]
 
-    if (!direct) {
-      const withJs = lastPart + '.js'
-      const withJsResult = lookupFn(pathToFile.concat(withJs), direct)
-      if (isResolveSuccess(withJsResult)) {
-        return withJsResult
+      // LOAD_AS_FILE(M)
+      const mainEntryResult = abstractLoadAsFile(fileForPathParts, mainEntryPathParts)
+      if (isResolveSuccess(mainEntryResult)) {
+        return mainEntryResult
       }
 
-      const withJsx = lastPart + '.jsx'
-      const withJsxResult = lookupFn(pathToFile.concat(withJsx), direct)
-      if (isResolveSuccess(withJsxResult)) {
-        return withJsxResult
+      // LOAD_INDEX(M)
+      const loadIndexMResult = loadIndex(fileForPathParts, mainEntryPathParts)
+      if (isResolveSuccess(loadIndexMResult)) {
+        return loadIndexMResult
       }
 
-      // TODO this also needs JSON parsing
-      const withJson = lastPart + '.json'
-      const withJsonResult = lookupFn(pathToFile.concat(withJson), direct)
-      if (isResolveSuccess(withJsonResult)) {
-        return withJsonResult
-      }
+      // LOAD_INDEX(X) DEPRECATED - note that this is only deprecated in the case where "main" is truthy
+      return loadIndex(fileForPathParts, normalisedPathParts)
     }
+  }
 
-    return resolveNotPresent
+  // If "main" is a falsy value
+  // LOAD_INDEX(X)
+  return loadIndex(fileForPathParts, normalisedPathParts)
+}
+
+// LOAD_INDEX(X)
+function loadIndex(fileForPathParts: FileForPathParts, pathParts: string[]): FileLookupResult {
+  // 1. If X/index.js is a file
+  //   a. Find the closest package scope SCOPE to X.
+  //   b. If no scope was found, load X/index.js as a CommonJS module. STOP.
+  //   c. If the SCOPE/package.json contains "type" field,
+  //     1. If the "type" field is "module", load X/index.js as an ECMAScript module. STOP.
+  //     2. Else, load X/index.js as an CommonJS module. STOP.
+  // Note: We don't attempt to load as an ECMAScript, as we can't - instead we'll try to load whatever
+  // it is as CommonJS, and if it turns out to be an esm we'll attempt to transpile it to CommonJS
```

Comment:
can you put the comments to the actual code lines?

Reply:
I left these comments here as a reminder that this is how node does this, but since we don't actually use the "type" field, we just load `X/index.js` as is, we don't need to look up the closest package scope here (hence the note at the bottom of the comment)